### ISSUE TITLE
TF2 to ONNX - Resnet Pretrained Model - Output Names changed to Identity

### ISSUE THREAD

**Issue by sampathweb** [onnx/tensorflow-onnx#921]
**Describe the bug**
I am trying to convert a TF 2 Saved Model to ONNX.  The Conversion runs fine. But the output node names get changed to Identity.  This may be OK for this model, but in another model, I have two outputs and I get output names as Identity and Identity_1 and it comes in different order in the outputs so the output names are important to get the correct outputs.

**Urgency**
If there are particular important use cases blocked by this or strict project-related timelines, please share more information and dates. If there are no hard deadlines, please specify none.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Tensorflow Version: TF 2.1
- Python version: 3.6

**To Reproduce**
Describe steps/code to reproduce the behavior:

```python
import tensorflow as tf
model = tf.keras.applications.ResNet50()
tf.saved_model.save(model, "resnet")
```
TF Node Name:
```
$ saved_model_cli show --dir resnet/  --tag_set serve --signature_def serving_default
The given SavedModel SignatureDef contains the following input(s):
  inputs['input_1'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 224, 224, 3)
      name: serving_default_input_1:0
The given SavedModel SignatureDef contains the following output(s):
  outputs['predictions'] tensor_info:
      dtype: DT_FLOAT
      shape: (-1, 1000)
      name: StatefulPartitionedCall:0
Method name is: tensorflow/serving/predict
```

TF2ONNX:
```bash
python -m tf2onnx.convert --opset 11 --fold_const --saved-model resnet/ --output resnet.onnx
```
When I look at Graph output, the output node name is Identity
```
in [5]: onnx_model.model.graph.output
Out[5]: 
[name: "Identity:0"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_param: "unk__619"
      }
      dim {
        dim_value: 1000
      }
    }
  }
}
]
```

**Expected behavior**

Output node names should match TF Output Node name of "predictions" instead of "Identity".  But a Identity node gets added changing the output node name.

I tried specifying `--outputs` in `tf2onnx.convert` but it would still attach a Identity node.  How do I get the output names to match the TF Model?

**Comment by jignparm** [comment]
I noticed the same behavior for TF models generated from Keras, and converted to Onnx -- the output name ends up being 'Identity' -- which, despite producing correct  values, seems incorrect.

**Comment by guschmue** [comment]
we use ``` concrete_func.outputs``` to get the outputs. 
Wired that it is different than what the saved model has as output.
Let me see if I can switch it to use the outputs from saved-model or the ones from keras.

**Comment by guschmue** [comment]
ok, going to fix this shortly.

**Comment by guschmue** [comment]
I think I promised to early. Tried that structured_output but that is the same thing.
Even if I do something like:
```
model.add(layers.Lambda(lambda x: x, name="new_layer_name"))
model.output
# <tf.Tensor 'new_layer_name/Identity:0' shape=(None, 1) dtype=float32>
```
The saved_model has
```
signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['dense_6_input'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 9)
        name: serving_default_dense_6_input:0
  The given SavedModel SignatureDef contains the following output(s):
    outputs['new_layer_name'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 1)
        name: StatefulPartitionedCall:0
```

**Comment by grebdioZ** [comment]
Any new ideas on this? We encountered the same thing when trying to convert our TF model repo to ONNX, and at current it looks like we would have to change all the output name references.

**Comment by hmeine** [comment]
I tried to reproduce this.

In http://netron.app, I uploaded the resulting resnet.onnx and could see:
* The second-to-last block in the visualization was a SoftMax node; when I select this, the "NODE PROPERTIES" on the right show me that its name is `StatefulPartitionedCall/resnet50/predictions/Softmax`, which looks reasonably good (contains "predictions").
* The last block in the visualization is labeled "Identity:0"; when I select this, I get "MODEL PROPERTIES" on the right (global, same as when selecting the first / input block).

Apparently, the output *tensor* has the wrong (unexpected) name, but the last *node* is correctly named.

Interestingly, this is the output from tf2onnx:

```
2021-01-06 12:08:28,929 - WARNING - '--tag' not specified for saved_model. Using --tag serve
2021-01-06 12:08:37,126 - INFO - Signatures found in model: [serving_default].
2021-01-06 12:08:37,126 - WARNING - '--signature_def' not specified, using first signature: serving_default
2021-01-06 12:08:42,486 - WARNING - From /Users/hmeine/Library/Python/3.7/lib/python/site-packages/tf2onnx/tf_loader.py:416: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.compat.v1.graph_util.extract_sub_graph`
2021-01-06 12:08:46,136 - INFO - Using tensorflow=2.3.1, onnx=1.6.0, tf2onnx=1.7.2/995bd6
2021-01-06 12:08:46,137 - INFO - Using opset <onnx, 11>
2021-01-06 12:08:49,418 - INFO - Computed 0 values for constant folding
2021-01-06 12:08:59,214 - INFO - Optimizing ONNX model
2021-01-06 12:09:09,326 - INFO - After optimization: Add -1 (18->17), BatchNormalization -53 (53->0), Cast -1 (1->0), Const -162 (271->109), Identity -57 (57->0), Reshape -1 (1->0), Transpose -213 (215->2)
2021-01-06 12:09:09,528 - INFO -
2021-01-06 12:09:09,528 - INFO - Successfully converted TensorFlow model resnet/ to ONNX
2021-01-06 12:09:09,690 - INFO - ONNX model is saved at resnet.onnx
```

This tells me that during network optimization, **57 Identity nodes were removed**. Could one of these have been behind the softmax and explain the tensor name?

**Comment by guschmue** [comment]
tensorflow adds those Identity nodes when we ask for the graph, even in tensorflow itself you'll see those names.
I have not found away around this yet.
Maybe we can do some thing to change the names to what the saved model says they should be.

**Comment by hmeine** [comment]
What exactly was the problem with `structured_outputs`?
```
In [1]: import tensorflow as tf

In [2]: model = tf.saved_model.load('resnet')

In [3]: sig = model.signatures['serving_default']

In [4]: sig.outputs
Out[4]: [<tf.Tensor 'Identity:0' shape=(None, 1000) dtype=float32>]

In [5]: sig.structured_outputs
Out[5]: {'predictions': TensorSpec(shape=(None, 1000), dtype=tf.float32, name='predictions')}
```

**Comment by guschmue** [comment]
The problem is that the graph that tensorflow returns is using the identity nodes.

**Comment by TomWildenhain-Microsoft** [comment]
I think using the names provided by structured outputs would be ideal and should be possible if we rename the identity nodes to match those names when we load the graph.  To avoid breaking existing conversions we should probably put it behind a flag.

@hmeine In the meantime yes, unfortunately you need to guess the output order and use the suboptimal names "Identity", "Identity_1" etc.  You can also manually edit the graph after conversion to use your preferred names.  To do so, you might find the experimental onnx2py tool useful: https://github.com/microsoft/onnxconverter-common/blob/master/onnxconverter_common/onnx2py.py

**Comment by hmeine** [comment]
Using the names from structured outputs would be great indeed!

I am not even sure if anybody would complain about "broken existing conversions", but you probably know the average tf2onnx user better than me. In my opinion, anybody actually caring about the names should see this as a bugfix. But of course it's a behavior change, so I see where you're coming from.

**Comment by guschmue** [comment]
We'll see that we rename the identity nodes to the structured output names.

**Comment by grebdioZ** [comment]
That sounds great, thank you very much for looking into it again. Do you already have a coarse idea of the timeline for a patch?

**Comment by grebdioZ** [comment]
> Pretty high on the list, hope to have this done next week.

Perfect, thanks!

**Comment by Baylibrejmonnet** [comment]
hi, glad to see, that i'm not the only one, who noticed this issue .

what about a patch ? 
thank you !

**Comment by TomWildenhain-Microsoft** [comment]
If you convert using the latest tf2onnx from master, output names for saved models will now use structured outputs.  You can also rename outputs with the --rename-outputs flag (and provide a list of desired names).
